File: src/context.rs
use block2::RcBlock;
use objc2::rc::Retained;
use objc2_app_kit::{
    NSApplicationActivationOptions, NSRunningApplication, NSWorkspace,
    NSWorkspaceDidActivateApplicationNotification,
};
use objc2_foundation::{NSNotification, NSOperationQueue};
use std::env;
use std::process;
use std::ptr::NonNull;
use std::rc::Rc;
use sysinfo::{Pid, System};

#[derive(Clone)]
pub struct ProcessContext {
    pub bundle_id: Option<String>,
    pub app_name: Option<String>,
    pub app_pid: Option<u32>,
}

impl ProcessContext {
    pub fn start_focus_monitoring(&self) {
        if let Some(target_pid) = self.app_pid {
            unsafe {
                let workspace = Rc::new(NSWorkspace::sharedWorkspace());
                let workspace_ref = workspace.clone();
                let check_and_exit_if_focused = move || {
                    // When notified, check if the frontmost app is our target
                    if let Some(front) = workspace_ref.frontmostApplication() {
                        if front.processIdentifier() as u32 == target_pid {
                            std::process::exit(0);
                        }
                    }
                };

                check_and_exit_if_focused();

                let notification_center = workspace.notificationCenter();
                let queue = NSOperationQueue::mainQueue();

                // Monitor application activation events
                let block = RcBlock::new(move |_: NonNull<NSNotification>| {
                    check_and_exit_if_focused();
                });

                let block_ptr: &block2::Block<dyn Fn(NonNull<NSNotification>)> = &block;

                notification_center.addObserverForName_object_queue_usingBlock(
                    Some(NSWorkspaceDidActivateApplicationNotification),
                    None,
                    Some(&queue),
                    block_ptr,
                );

                // Leak the block to keep it alive for the duration of the program
                Box::leak(Box::new(block));
            }
        }
    }

    pub fn current() -> Self {
        let mut sys = System::new_all();
        sys.refresh_all();

        let current_pid = Pid::from_u32(process::id());

        let parent_pid = sys
            .process(current_pid)
            .and_then(|p| p.parent())
            .map(|p| p.as_u32());

        let mut context = ProcessContext {
            bundle_id: None,
            app_name: None,
            app_pid: None,
        };

        if let Some(ppid) = parent_pid {
            context.resolve_app_info(&sys, Pid::from_u32(ppid));
        }

        context
    }

    fn resolve_app_info(&mut self, sys: &System, ppid: Pid) {
        // 首先尝试从环境变量获取终端信息
        if let Some((app_name, bundle_id, pid)) = get_terminal_from_env() {
            self.app_name = Some(app_name);
            self.bundle_id = Some(bundle_id);
            self.app_pid = Some(pid);
            return;
        }

        let mut current_pid = ppid;

        while let Some(proc) = sys.process(current_pid) {
            let pid_u32 = current_pid.as_u32();
            let proc_name = proc.name();

            if let Some(bundle_id) = get_bundle_id(pid_u32) {
                self.bundle_id = Some(bundle_id);
                self.app_name = Some(proc_name.to_string());
                self.app_pid = Some(pid_u32);

                if !proc_name.to_lowercase().contains("helper")
                    && !proc_name.to_lowercase().contains("zsh")
                    && !proc_name.to_lowercase().contains("bash")
                    && !proc_name.to_lowercase().contains("login")
                {
                    break;
                }
            }

            if let Some(next_ppid) = proc.parent() {
                current_pid = next_ppid;
            } else {
                break;
            }
        }

        // Fallback for Ghostty or other terminals that break the process tree
        if self.app_pid.is_none() {
            for proc in sys.processes().values() {
                let name = proc.name().to_lowercase();
                if name.contains("ghostty") || name.contains("zed") {
                    let pid = proc.pid().as_u32();
                    if let Some(bid) = get_bundle_id(pid) {
                        self.bundle_id = Some(bid);
                        self.app_name = Some(proc.name().to_string());
                        self.app_pid = Some(pid);
                        break;
                    }
                }
            }
        }
    }

    pub fn activate(&self) -> bool {
        if let Some(pid) = self.app_pid {
            // Try AppleScript first for specific window activation
            if self.activate_via_applescript(pid) {
                return true;
            }

            unsafe {
                let app = NSRunningApplication::runningApplicationWithProcessIdentifier(pid as i32);
                if let Some(app) = app {
                    #[allow(deprecated)]
                    return app.activateWithOptions(
                        NSApplicationActivationOptions::NSApplicationActivateIgnoringOtherApps,
                    );
                }
            }
        }
        false
    }

    fn activate_via_applescript(&self, pid: u32) -> bool {
        let current_dir = env::current_dir().ok();
        let dir_name =
            current_dir.and_then(|p| p.file_name().map(|n| n.to_string_lossy().to_string()));

        if let Some(name) = dir_name {
            let script = format!(
                r#"
                tell application "System Events"
                    try
                        set targetProc to first process whose unix id is {}
                        tell targetProc
                            set targetWin to (first window whose name contains "{}")
                            perform action "AXRaise" of targetWin
                            set frontmost to true
                            return "true"
                        end tell
                    on error
                        return "false"
                    end try
                end tell
                "#,
                pid,
                name.replace("\"", "\\\"")
            );

            let output = process::Command::new("osascript")
                .arg("-e")
                .arg(script)
                .output();

            if let Ok(out) = output {
                let stdout = String::from_utf8_lossy(&out.stdout).trim().to_string();
                return stdout == "true";
            }
        }
        false
    }
}

fn get_bundle_id(pid: u32) -> Option<String> {
    unsafe {
        let app: Option<Retained<NSRunningApplication>> =
            NSRunningApplication::runningApplicationWithProcessIdentifier(pid as i32);
        app.and_then(|a| a.bundleIdentifier())
            .map(|s| s.to_string())
    }
}

/// 从环境变量获取终端信息
/// 读取 TERM_PROGRAM 环境变量来识别终端类型
fn get_terminal_from_env() -> Option<(String, String, u32)> {
    let term_program = std::env::var("TERM_PROGRAM").ok()?;

    let (bundle_id, app_name) = match term_program.to_lowercase().as_str() {
        "zed" => ("dev.zed.Zed", "Zed"),
        "ghostty" => ("com.mitchellh.ghostty", "Ghostty"),
        "Apple_Terminal" => ("com.apple.Terminal", "Terminal"),
        "iTerm.app" => ("com.googlecode.iterm2", "iTerm"),
        _ => return None,
    };

    // 获取终端进程的 PID
    if let Some(pid) = find_terminal_pid(bundle_id) {
        Some((app_name.to_string(), bundle_id.to_string(), pid))
    } else {
        None
    }
}

/// 通过 bundle ID 查找终端进程的 PID
fn find_terminal_pid(bundle_id: &str) -> Option<u32> {
    unsafe {
        let workspace = NSWorkspace::sharedWorkspace();
        let apps = workspace.runningApplications();
        // 使用索引遍历 NSArray
        let count = apps.count();
        for i in 0..count {
            let app = apps.objectAtIndex(i);
            if let Some(bid) = app.bundleIdentifier() {
                if bid.to_string() == bundle_id {
                    return Some(app.processIdentifier() as u32);
                }
            }
        }
        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_current_context() {
        let context = ProcessContext::current();
        // Just verify it doesn't crash and we get some info
        assert!(context.app_name.is_some() || context.bundle_id.is_some());
    }
}

File: terminal-notificator-swift/Sources/TerminalNotificator/Context.swift
import Foundation
import AppKit

struct TerminalContext {
    let bundleId: String
    let appName: String
    let appPid: pid_t
    let currentDirectory: String

    /// 是否支持 OSC 9 通知（如 Ghostty）
    var supportsOSC9: Bool {
        return bundleId == "com.mitchellh.ghostty"
    }

    @MainActor
    static func detect() async throws -> TerminalContext {
        // 1. 尝试从环境变量获取
        if let context = await getTerminalFromEnv() {
            return context
        }

        // 2. 如果环境变量没有，通过进程树查找
        let ppid = getppid()
        if let context = try await resolveAppInfo(startingPid: ppid) {
            return context
        }

        throw NotificationError.terminalContextNotFound
    }

    @MainActor
    func activate() async -> Bool {
        // 先尝试用 AppleScript 激活特定窗口
        if await activateViaAppleScript(pid: appPid, currentDirectory: currentDirectory) {
            return true
        }

        // 回退到用 NSWorkspace 激活整个应用
        if let app = NSRunningApplication(processIdentifier: appPid) {
            return app.activate(options: [])
        }

        return false
    }
}

// MARK: - Focus Check

extension TerminalContext {
    @MainActor
    func isFrontmostAndActive() async -> Bool {
        guard let frontmostApp = NSWorkspace.shared.frontmostApplication else {
            return false
        }

        guard frontmostApp.bundleIdentifier == bundleId,
              frontmostApp.processIdentifier == appPid else {
            return false
        }

        // 对于某些终端（如 Ghostty），窗口标题是命令而非目录
        // 如果应用本身在焦点，就认为终端在焦点
        let terminalsToSkipWindowCheck = ["com.mitchellh.ghostty"]
        if terminalsToSkipWindowCheck.contains(bundleId) {
            return true
        }

        // 对于其他终端，检查窗口是否匹配目录
        return await checkFrontmostWindowMatches(currentDirectory: currentDirectory)
    }
}

// MARK: - Helper Functions

@MainActor
private func getTerminalFromEnv() async -> TerminalContext? {
    guard let termProgram = ProcessInfo.processInfo.environment["TERM_PROGRAM"] else {
        return nil
    }

    let mapping: [String: (String, String)] = [
        "zed": ("dev.zed.Zed", "Zed"),
        "ghostty": ("com.mitchellh.ghostty", "Ghostty"),
        "apple_terminal": ("com.apple.Terminal", "Terminal"),
        "iterm.app": ("com.googlecode.iterm2", "iTerm")
    ]

    guard let (bundleId, appName) = mapping[termProgram.lowercased()] else {
        return nil
    }

    if let pid = await findTerminalPid(bundleId: bundleId) {
        let currentDir = FileManager.default.currentDirectoryPath
        return TerminalContext(bundleId: bundleId, appName: appName, appPid: pid, currentDirectory: currentDir)
    }

    return nil
}

@MainActor
private func findTerminalPid(bundleId: String) async -> pid_t? {
    let workspace = NSWorkspace.shared
    let apps = workspace.runningApplications
    for app in apps {
        if app.bundleIdentifier == bundleId {
            return app.processIdentifier
        }
    }
    return nil
}

@MainActor
private func resolveAppInfo(startingPid: pid_t) async throws -> TerminalContext? {
    var currentPid = startingPid
    let currentDir = FileManager.default.currentDirectoryPath

    while currentPid > 1 {
        if let app = NSRunningApplication(processIdentifier: currentPid) {
            let procName = app.localizedName ?? ""
            let lowerName = procName.lowercased()

            if let bundleId = app.bundleIdentifier {
                if !lowerName.contains("helper") &&
                   !lowerName.contains("zsh") &&
                   !lowerName.contains("bash") &&
                   !lowerName.contains("login") {
                    return TerminalContext(bundleId: bundleId, appName: procName, appPid: currentPid, currentDirectory: currentDir)
                }
            }
        }

        var mib = [CTL_KERN, KERN_PROC, KERN_PROC_PID, currentPid]
        var info = kinfo_proc()
        var size = MemoryLayout<kinfo_proc>.stride

        let result = sysctl(&mib, u_int(mib.count), &info, &size, nil, 0)
        if result == 0 {
            currentPid = info.kp_eproc.e_ppid
        } else {
            break
        }
    }

    return nil
}

// MARK: - AppleScript Helper

/// 执行 AppleScript 并返回输出结果
private func runAppleScript(_ script: String) -> String? {
    let task = Process()
    task.launchPath = "/usr/bin/osascript"
    task.arguments = ["-e", script]

    let pipe = Pipe()
    task.standardOutput = pipe

    do {
        try task.run()
        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        return String(data: data, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines)
    } catch {
        return nil
    }
}

private func activateViaAppleScript(pid: pid_t, currentDirectory: String) async -> Bool {
    let dirName = (currentDirectory as NSString).lastPathComponent

    let script = """
    tell application "System Events"
        try
            set targetProc to first process whose unix id is \(pid)
            tell targetProc
                -- 首先尝试匹配完整目录路径
                set targetWin to missing value
                try
                    set targetWin to (first window whose name contains "\(currentDirectory)")
                on error
                    try
                        -- 回退到只匹配目录名
                        set targetWin to (first window whose name contains "\(dirName)")
                    on error
                        -- 如果都找不到，使用第一个窗口
                        set targetWin to first window
                    end try
                end try

                if targetWin is not missing value then
                    perform action "AXRaise" of targetWin
                    set frontmost to true
                    return "true"
                end if
                return "false"
            end tell
        on error
            return "false"
        end try
    end tell
    """

    return runAppleScript(script) == "true"
}

@MainActor
private func checkFrontmostWindowMatches(currentDirectory: String) async -> Bool {
    let dirName = (currentDirectory as NSString).lastPathComponent

    let script = """
    tell application "System Events"
        try
            set frontApp to first application process whose frontmost is true
            tell frontApp
                try
                    set frontWin to first window
                    return name of frontWin
                on error
                    return ""
                end try
            end tell
        on error
            return ""
        end try
    end tell
    """

    guard let winName = runAppleScript(script), !winName.isEmpty else {
        return false
    }

    return winName.contains(currentDirectory) || winName.contains(dirName)
}
